File: 590.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/
int x = []() {
  std::ios::sync_with_stdio(false);
  std::cin.tie(nullptr);
  return 0;
}();

class Solution {
public:
    vector<int> postorder(Node* root) {
     	vector<int> res;
     	if (!root)
     	{
     		return res;
     	}
     	stack<Node*> s;
        s.push(root);
     	Node* tem = root;
     	while(!s.empty()) {
     		if (!is_a_father_of_b(s.top(), tem))
     		{
     			go(s);
     		}
     		tem = s.top();
     		s.pop();
     		res.push_back(tem->val);
     	}
     	return res;
    }

    void go(stack<Node*> &s) {
    	while (1){
    		Node* tem = s.top();
    		if (!tem) break;
    		else {
    			auto c = tem->children;
    			if(!c.empty()) {
    				for (int i = c.size() - 1; i >= 0; --i){
    					s.push(c[i]);
    				}
    			}
    			else{
    				break;
    			}
    		}
    	}
	}

    bool is_a_father_of_b(Node* a, Node* b) {
    	for(Node* t : a->children) {
    		if (t == b) return true;
    	}
    	return false;
    }
};

File: 590.go
package leetcode

//  Definition for a Node.
type Node struct {
	Val      int
	Children []*Node
}

func postorder(root *Node) []int {
	ans := []int{}
	if root == nil {
		return ans
	}

	stack := []*Node{root}
	for len(stack) > 0 {
		node := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		ans = append(ans, node.Val)
		stack = append(stack, node.Children...)
	}

	for i, j := 0, len(ans)-1; i < j; i++ {
		ans[i], ans[j] = ans[j], ans[i]
		j--
	}

	return ans
}

func postorder2(root *Node) (ans []int) {
	if root == nil {
		return
	}
	stack := []interface{}{root}
	for l := len(stack); l > 0; l = len(stack) {
		obj := stack[l-1]
		stack = stack[:l-1]
		if value, ok := obj.(int); ok {
			ans = append(ans, value)
		} else {
			value, _ := obj.(*Node)
			stack = append(stack, value.Val)
			// 这个写法有点秀
			// 对于一个节点存在未访问、访问了但还没访问子节点、访问完子节点又访问三种状态
			// 在第三种状态的时候才可以把节点值放到答案里
			// 别的做法都是专门添加一个变量表示中间状态
			// 比如{node, isVisited}这种
			// 但是这种做法就很妙，直接把节点值当作中间状态
			for i := len(value.Children) - 1; i >= 0; i-- {
				stack = append(stack, value.Children[i])
			}
		}
	}
	return
}
